From 8f9e3bff3551ace00bdb985cc580aeb7af4c86f8 Mon Sep 17 00:00:00 2001
From: Yelena Krivosheev <yelena@marvell.com>
Date: Mon, 25 Dec 2017 15:28:01 +0200
Subject: [PATCH 2129/2241] fix: net: mvneta: prevent phy access before mac
 configuration

PHY can be connected to an interface only after full MAC
configuration. Otherwise we cannot configure MAC according
to PHY autoneg results.

Currently PHY is connected to port with mvneta_probe(), before
opening the port.
This patch fixes this to be in the correct order, by moving:
mvneta_mdio_probe()  -> mvneta_open() instead mvneta_probe()
mvneta_mdio_remove() -> mvneta_stop() instead mvneta_remove()

Change-Id: Ie517d41e6ef94895af56de18c7d5fbeb1760761d
Signed-off-by: Yelena Krivosheev <yelena@marvell.com>
Reviewed-on: http://vgitil04.il.marvell.com:8080/48106
Reviewed-by: Omri Itach <omrii@marvell.com>
Tested-by: Omri Itach <omrii@marvell.com>
---
 drivers/net/ethernet/marvell/mvneta.c | 20 +++++++++++---------
 1 file changed, 11 insertions(+), 9 deletions(-)

diff --git a/drivers/net/ethernet/marvell/mvneta.c b/drivers/net/ethernet/marvell/mvneta.c
index a9ce397..6db1469 100644
--- a/drivers/net/ethernet/marvell/mvneta.c
+++ b/drivers/net/ethernet/marvell/mvneta.c
@@ -3796,6 +3796,14 @@ static int mvneta_open(struct net_device *dev)
 	/* In default link is down */
 	netif_carrier_off(pp->dev);
 
+	if (!pp->use_inband_status) {
+		ret = mvneta_mdio_probe(pp);
+		if (ret < 0) {
+			netdev_err(dev, "cannot probe MDIO bus\n");
+			return 0;
+		}
+	}
+
 	mvneta_start_dev(pp);
 
 	return 0;
@@ -3831,6 +3839,9 @@ static int mvneta_stop(struct net_device *dev)
 		free_irq(dev->irq, pp);
 	}
 
+	if (!pp->use_inband_status)
+		mvneta_mdio_remove(pp);
+
 	mvneta_cleanup_rxqs(pp);
 	mvneta_cleanup_txqs(pp);
 
@@ -4875,13 +4886,6 @@ static int mvneta_probe(struct platform_device *pdev)
 	if (pp->flags & MVNETA_PORT_F_IF_MUSDK)
 		netdev_info(dev, "Port belong to User Space (MUSDK)\n");
 
-	if (!pp->use_inband_status) {
-		err = mvneta_mdio_probe(pp);
-		if (err < 0) {
-			netdev_err(dev, "cannot probe MDIO bus\n");
-			goto err_netdev;
-		}
-	}
 	return 0;
 
 err_netdev:
@@ -4918,8 +4922,6 @@ static int mvneta_remove(struct platform_device *pdev)
 	struct net_device  *dev = platform_get_drvdata(pdev);
 	struct mvneta_port *pp = netdev_priv(dev);
 
-	if (!pp->use_inband_status)
-		mvneta_mdio_remove(pp);
 	unregister_netdev(dev);
 	clk_disable_unprepare(pp->clk);
 	free_percpu(pp->ports);
-- 
2.7.4

